network/benchmark/src/lib.rs

```diff
@@ -0,0 +1,605 @@
+            };
+        }
+        let mut oldest = self.sent[0].send_micros;
+        let capacity = self.sent.len();
+        for i in 0..capacity {
+            let pos = (self.sent_pos + capacity - (1 + i)) % capacity;
+            let rec = self.sent[pos].clone();
+            if rec.request_counter == request_counter {
+                return rec;
+            }
+            if rec.send_micros < oldest {
+                oldest = rec.send_micros;
+            }
+        }
+        SendRecord {
+            request_counter: 0,
+            send_micros: oldest,
+            bytes_sent: 0,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct SendRecord {
+    pub request_counter: u64,
+    pub send_micros: i64,
+    pub bytes_sent: usize,
+}
+
+pub static APTOS_NETWORK_BENCHMARK_DIRECT_MESSAGES: Lazy<IntCounterVec> = Lazy::new(|| {
+    register_int_counter_vec!(
+        "aptos_network_benchmark_direct_messages",
+        "Number of benchmark direct messages",
+        &["state"]
+    )
+    .unwrap()
+});
+
+fn direct_messages(state_label: &'static str) {
+    APTOS_NETWORK_BENCHMARK_DIRECT_MESSAGES
+        .with_label_values(&[state_label])
+        .inc();
+}
+
+pub static APTOS_NETWORK_BENCHMARK_DIRECT_BYTES: Lazy<IntCounterVec> = Lazy::new(|| {
+    register_int_counter_vec!(
+        "aptos_network_benchmark_direct_bytes",
+        "Number of benchmark direct bytes",
+        &["state"]
+    )
+    .unwrap()
+});
+
+fn direct_bytes(state_label: &'static str, byte_count: u64) {
+    APTOS_NETWORK_BENCHMARK_DIRECT_BYTES
+        .with_label_values(&[state_label])
+        .inc_by(byte_count);
+}
+
+pub static APTOS_NETWORK_BENCHMARK_DIRECT_MICROS: Lazy<IntCounterVec> = Lazy::new(|| {
+    register_int_counter_vec!(
+        "aptos_network_benchmark_direct_micros",
+        "Number of benchmark direct micros",
+        &["state"]
+    )
+    .unwrap()
+});
+
+fn direct_micros(state_label: &'static str, micros: u64) {
+    APTOS_NETWORK_BENCHMARK_DIRECT_MICROS
+        .with_label_values(&[state_label])
+        .inc_by(micros);
+}
+
+pub static APTOS_NETWORK_BENCHMARK_RPC_MESSAGES: Lazy<IntCounterVec> = Lazy::new(|| {
+    register_int_counter_vec!(
+        "aptos_network_benchmark_rpc_messages",
+        "Number of benchmark RPC messages",
+        &["state"]
+    )
+    .unwrap()
+});
+
+fn rpc_messages(state_label: &'static str) {
+    APTOS_NETWORK_BENCHMARK_RPC_MESSAGES
+        .with_label_values(&[state_label])
+        .inc();
+}
+
+pub static APTOS_NETWORK_BENCHMARK_RPC_BYTES: Lazy<IntCounterVec> = Lazy::new(|| {
+    register_int_counter_vec!(
+        "aptos_network_benchmark_rpc_bytes",
+        "Number of benchmark RPC bytes transferred",
+        &["state"]
+    )
+    .unwrap()
+});
+
+pub fn rpc_bytes(state_label: &'static str) -> IntCounter {
+    APTOS_NETWORK_BENCHMARK_RPC_BYTES.with_label_values(&[state_label])
+}
+
+pub static APTOS_NETWORK_BENCHMARK_RPC_MICROS: Lazy<IntCounterVec> = Lazy::new(|| {
```

Comment:
nit: why not move these into a `metrics.rs` or `counters.rs` file?

Reply:
Why not keep them where they're used?

Reply:
By that logic everything could go into one file? 😄 Also, I'm in favor of code consistency. If you look at every other application, we tend to put metrics into a separate file (e.g., `counters.rs` and `metrics.rs`). It just makes it easier when code starts to grow.